prometheus/promhttp/http.go

```diff
@@ -155,29 +145,50 @@ func HandlerFor(reg prometheus.Gatherer, opts HandlerOpts) http.Handler {
 				}
 			}
 		}
-		if closer, ok := writer.(io.Closer); ok {
-			closer.Close()
-		}
+
 		if lastErr != nil && buf.Len() == 0 {
 			http.Error(w, "No metrics encoded, last error:\n\n"+lastErr.Error(), http.StatusInternalServerError)
 			return
 		}
-		header := w.Header()
-		header.Set(contentTypeHeader, string(contentType))
-		header.Set(contentLengthHeader, fmt.Sprint(buf.Len()))
-		if encoding != "" {
-			header.Set(contentEncodingHeader, encoding)
-		}
+
+		w.Header().Set(contentTypeHeader, string(contentType))
+
 		if _, err := w.Write(buf.Bytes()); err != nil && opts.ErrorLog != nil {
 			opts.ErrorLog.Println("error while sending encoded metrics:", err)
 		}
 		// TODO(beorn7): Consider streaming serving of metrics.
 	})
 
+	gzipHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
It is helpful to have two `http.HandlerFunc`s here where one just wraps the other? Couldn't this just be inlined?

Reply:
Nested handlers is the normal way to implement gzip middleware in Go, especially if you want it to be optional / settable by some flag.

Reply:
I'm not a fan of it being an anonymous function either, as I suspect this will cause a lot of allocations / GC. The NYTimes implementation is worth studying. And bear in mind that the content-length header should be the length of the gzipped output, not the original input.

Reply:
WRT content length: Yes, I had that in mind. That's why I suggested it's probably OK to not add a content length when gzipping. But we should (or even SHOULD :-) still add the content length when _not_ gzipping. And that's not possible with the approach here where the decision is made within the gzipHandler local function but the uncompressed content length is only available in the h local function. In this specific case, where we don't want to provide a generically usable gzip handler, I'd probably just inline everything into one local function. At least for now. We can still refactor into something nicer once we have something that works.